alibi_detect/utils/schemas.py

```diff
@@ -1,32 +1,30 @@
 # type: ignore[assignment]
 # TODO - conditional checks depending on backend
 # TODO - consider validating output of get_config calls
-
+# TODO check https://pydantic-docs.helpmanual.io/usage/postponed_annotations/ - think OK but test
 import numpy as np
 from pydantic import BaseModel
 from typing import Optional, Union, Dict, List, Callable
 from alibi_detect.utils._types import Literal
 from alibi_detect.version import __version__, __config_spec__
-from alibi_detect.cd.tensorflow import HiddenOutput, UAE
 from alibi_detect.models.tensorflow import TransformerEmbedding
 from transformers import PreTrainedTokenizerBase
-from alibi_detect.utils.frameworks import has_pytorch, has_tensorflow  # , has_sklearn
-
+from alibi_detect.utils.frameworks import has_pytorch, has_tensorflow
 
-# Populate supported models
-# SupportedModels is a tuple of the real class types, to be used in isinstance() in saving.py and loading.py.
-# SUPPORTED_MODELS is a typing.Union of the same model types, to be used in type hints.
-SupportedModels = [UAE, HiddenOutput]
+# SupportedModels_types is a tuple of possible models (conditional on installed deps). This is used in isinstance() etc.
+SupportedModels = []
 if has_tensorflow:
     import tensorflow as tf
-    SupportedModels.append(tf.keras.Model)
+    from alibi_detect.cd.tensorflow import UAE, HiddenOutput
+    SupportedModels += [tf.keras.Model, UAE, HiddenOutput]
 if has_pytorch:
     import torch
 #    SupportedModels.append()  # TODO
 # if has_sklearn:
+#    import sklearn
 #    SupportedModels.append()  # TODO
 SupportedModels = tuple(SupportedModels)
-SUPPORTED_MODELS = Union[SupportedModels]
+SupportedModels_py = Union[SupportedModels]  # only for use with pydantic. NOT to be used with mypy (as not static)
```

Comment:
I've had to regress from the changes made in part II (see https://github.com/SeldonIO/alibi-detect/pull/462#discussion_r827852989). The part II solution didn't actually work properly when the `Union` was used in `saving/loading.py`. I suspect this is because we were trying to populate a `Union` with if-else statements, when mypy is a static type checker.

The new `SupportedModels_py` `Union` is constructed the same way, but is only used by pydantic. It seems using a dynamically generated `Union` is fine for that, but not for mypy. The type signatures used in `loading/saving.py` for models are relaxed to just be `Callable`....